lib/hacluster.pm

```diff
@@ -1248,4 +1250,95 @@ sub cluster_status_matches_regex {
     }
 }
 
+=head3 crm_maintenance_status
+
+    crm_maintenance_status();
+
+Check maintenance mode status. Returns true (maintenance active) or false (maintenance inactive).
+Croaks if unknown status is received.
+
+=over 1
+
+=back
+=cut
+
+sub crm_maintenance_status {
+    my $cmd_output = script_output('crm configure show cib-bootstrap-options | grep maintenance-mode');
+    $cmd_output =~ s/\s//g;
+    my $status = (split('=', $cmd_output))[-1];
+    croak "CRM returned unrecognized status: '$status'" unless grep(/^$status$/, ('false', 'true'));
+    return $status;
+}
+
+=head3 crm_wait_for_maintenance
+
+    crm_wait_for_maintenance(target_state=>$target_state, [loop_sleep=>$loop_sleep, timeout=>$timeout]);
+
+Wait for maintenance to be turned on or off. Croaks on timeout.
+
+B<target_state> Target state of the maintenance mode (true/false)
+B<loop_sleep> Override default sleep value between checks
+B<timeout> Override default timeout value
+
+=over 1
+
+=back
+=cut
+
+sub crm_wait_for_maintenance {
+    my (%args) = @_;
+    my $timeout = $args{timeout} // bmwqemu::scale_timeout(30);
+    my $target_state = $args{'target_state'};
+    my $loop_sleep = $args{loop_sleep} // 5;
+
+    croak "Invalid argument value: \$target_state = '$target_state'" unless grep(/^$target_state$/, ('false', 'true'));
+
+    my $current_status = crm_maintenance_status();
+    my $start_time = time;
+    while ($current_status ne $target_state) {
+        $current_status = crm_maintenance_status();
+        croak "Timeout while waiting for maintenance mode: '$target_state'" if (time - $start_time > $timeout);
+        sleep $loop_sleep;
+    }
+    record_info("Maintenance", "Maintenance status: $current_status");
+    return $current_status;
+}
+
+=head3 crm_check_resource_location
+
+    crm_check_resource_location(resource=>$resource, [wait_for_target=>$wait_for_target, timeout=>$timeout]);
+
+Checks current resource location, returns physical hostname of the node. Can be used to wait for desired state Eg: after failover.
+Croaks upon timeout.
+
+B<wait_for_target> Target location of the resource specified - physical hostname
+B<resource> Resource to check
+B<timeout> Override default timeout value
+
+=over 1
+
+=back
+=cut
+
+sub crm_check_resource_location {
+    my (%args) = @_;
+    my $resource = $args{resource};
+    my $wait_for_target = $args{wait_for_target} // 0;
+    my $timeout = $args{timeout} // bmwqemu::scale_timeout(120);
+    my $cmd = join(' ', "crm resource status", $resource, "| grep 'resource $resource is'");    # Grep to avoid random kernel message appearing in script_output
```

Comment:
```suggestion
    my $cmd = join(' ', "crm resource status", $args{resource}, "| grep 'resource $args{resource} is'");    # Grep to avoid random kernel message appearing in script_output
```

Reply:
changed